compiler/rustc_mir_transform/src/value_numbering.rs

```diff
@@ -0,0 +1,907 @@
+//! it's local to a basic block, however, it doesn't handle all the cases that our constant and copy
+//! propagation passes do. Eventually, when we have SSA, we should be able to upgrade this to global
+//! value numbering.
+
+use crate::simplify;
+use crate::ty::{Region, Ty};
+use crate::MirPass;
+use rustc_data_structures::fx::FxHashMap;
+use rustc_index::vec::IndexVec;
+use rustc_middle::mir::visit::{MutatingUseContext, PlaceContext, Visitor};
+use rustc_middle::mir::{
+    AggregateKind, BasicBlock, BasicBlockData, BinOp, Body, BorrowKind, CastKind, Constant,
+    InlineAsmOperand, Local, LocalDecls, Location, Mutability, NullOp, Operand, Place, PlaceElem,
+    ProjectionElem, Rvalue, SourceInfo, Statement, StatementKind, TerminatorKind, UnOp,
+};
+use rustc_middle::ty::{self, Const, ParamEnv, TyCtxt};
+use rustc_span::def_id::DefId;
+use std::collections::hash_map::Entry;
+use std::mem;
+
+/// Performs copy and constant propagation via value numbering.
+pub struct ValueNumbering;
+
+impl<'tcx> MirPass<'tcx> for ValueNumbering {
+    fn is_enabled(&self, sess: &rustc_session::Session) -> bool {
+        // We run on MIR opt level 2 or greater, if `-Z number-values` is passed.
+        //
+        // FIXME(pcwalton): Enable by default.
+        sess.mir_opt_level() >= 2 && sess.opts.debugging_opts.number_values
+    }
+
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, body: &mut Body<'tcx>) {
+        // First, promote mutable locals to immutable ones. This helps our rudimentary alias
+        // analysis a lot.
+        //
+        // FIXME(pcwalton): Should the pass manager be doing this?
+        simplify::promote_mutable_locals_to_immutable(body, tcx);
+
+        // Now loop over and optimize all basic blocks.
+        let param_env = tcx.param_env(body.source.def_id());
+        let (basic_blocks, local_decls) = body.basic_blocks_and_local_decls_mut();
+        for block_index in 0..basic_blocks.len() {
+            let block_index = BasicBlock::from_usize(block_index);
+            if basic_blocks[block_index].statements.is_empty() {
+                // This is necessary to avoid indexing underflows in `update_storage_dead_markers`,
+                // and it's probably faster anyway.
+                continue;
+            }
+
+            let mut numberer = ValueNumberer::new(tcx, param_env, block_index);
+            numberer.init_locals(block_index, &mut basic_blocks[block_index]);
+            numberer.process_statements(&mut basic_blocks[block_index], local_decls);
+            numberer.process_terminator(&mut basic_blocks[block_index]);
+            numberer.update_storage_dead_markers(block_index, basic_blocks);
+        }
+
+        // Finally, prune dead locals, as value numbering frequently makes dead stores.
+        //
+        // FIXME(pcwalton): As above, should the pass manager be doing this?
+        simplify::remove_unused_locals(body, tcx);
```

Comment:
(Yes)

My idea was to allow each pass to schedule any cleanup passes it needs, while allowing later passes to defer certain cleanup passes if a) they would benefit from that cleanup and b) they wouldn't be affected by running on non-cleaned up MIR. This would avoid the need to manually place `o1(SimplifyCfg)` in the pass manager while still allowing you to dump MIR prior to cleanup, etc. It also might be overkill, so I've been putting off implementing it.